### PR TITLE
fix: binary execution mode compatibility 

### PR DESCRIPTION
**Pull request by negar-abbasi**
- [#402](https://github.com/algorandfoundation/algokit-cli/issues/402)
- [#407](https://github.com/algorandfoundation/algokit-cli/issues/407)

### 402 
  - Implementing `multiprocessing.freeze_support()` to prevent execution failures when running `algokit task vanity-address` in binary.
 - Addressing the issue where algokit init fails during the execution of copier tasks in the fullstack template. The proposed solution involves using the system's Python path as a temporary substitute for `_copier_python`. This change is temporary until the issue with copier is resolved.

### 407
- Adding portability tests against `doctor`, `init`, `vanity-address` and `help` commands. 
- Minor bumps in versions of several github actions (setup-python to v5, download-artifacts to v4 and etc). Fixes several minor deprecation warnings in github workers logs
 
## Further notes

- Make sure to merge [fullstack](https://github.com/algorandfoundation/algokit-fullstack-template) PR first (marking it as a breaking change with a bump in min algokit version requirement)

### CODE REVIEW

File: src/algokit/cli/init.py [hunk 1]
```diff
@@ -216,6 +217,15 @@ def init_command(  # noqa: PLR0913
     # provide the directory name as an answer to the template, if not explicitly overridden by user
     answers_dict.setdefault("project_name", directory_name)
 
+    system_python_path = next(get_python_paths(), None)
+    if system_python_path is not None:
```
Comment (aorumbayev) [inline src/algokit/cli/init.py:221]: @achidlow @negar-abbasi - what if we expand this and also add usage of `is_binary_mode` and if its TRUE we execute a python script using something like [this](https://github.com/pyinstaller/pyinstaller/issues/2490) via `exec` i believe it should refer to currently used interpreter which in case of pyinstaller will be the one bundled along with the bootloader. That way users of binary installed algokit won't need to have python to run custom tasks in copier

Comment (achidlow) [reply src/algokit/cli/init.py:221]: I think the solution in this PR is simpler. Adding another alogkit command which is called from a copier template `_tasks` to execute a Python script that is part of template's data is a very tight coupling compared to informing a template of a Python interpreter location on the `$PATH` via a supplied answer.

File: src/algokit/__main__.py [hunk 2]
```diff
@@ -1,3 +1,10 @@
+from multiprocessing import freeze_support
+
 from algokit.cli import algokit
+from algokit.cli.common.utils import is_binary_mode
+
+# Required to support full feature parity when running in binary execution mode
+if is_binary_mode():
```
Comment (achidlow) [inline src/algokit/__main__.py:7]: It seems from the docs we can omit this `if` check and just always run `freeze_supprt`? https://docs.python.org/3/library/multiprocessing.html#multiprocessing.freeze_support

File: src/algokit/cli/init.py (hunk 1 above)
Comment (aorumbayev) [reply src/algokit/cli/init.py:221]: @achidlow i see so this is all handled by `get_python_paths`, if `which` did not return `python` or `python3` it will grab the `sys.executable` which will point to the interpreter bundled in binary (when running in binary mode).

**Comment by aorumbayev** [comment]
Hey @negar-abbasi, i figured i'll combine this with portability tests issue (will update pr description later), still need a bit more time to refactor the pipelines a bit, i have wip progress on a fork will finalize this pr tomorrow for @neilcampbell and yourself to review. In the meantime feel free to proceed with other items

File: src/algokit/cli/init.py [hunk 3]
```diff
@@ -216,6 +217,12 @@ def init_command(  # noqa: PLR0913
     # provide the directory name as an answer to the template, if not explicitly overridden by user
     answers_dict.setdefault("project_name", directory_name)
 
+    system_python_path = next(get_python_paths(), None)
```
Comment (neilcampbell) [inline src/algokit/cli/init.py:220]: Should we do the same for the generators? It's possible a generator might try to execute a python script via copier.

File: tests/init/test_init.py [hunk 4]
```diff
@@ -685,6 +707,57 @@ def test_init_with_custom_env(tmp_path_factory: TempPathFactory) -> None:
     )
 
 
+def test_init_fullstack_template_fails_on_missing_python(
```
Comment (neilcampbell) [inline tests/init/test_init.py:710]: This isn't technically the fullstack template.

You could also add a comment to indicate that the fullstack template is an example of a template which executes python as part of running copier.

```suggestion
def test_init_template_with_python_fails_on_missing_python(
```

File: tests/init/test_init.py [hunk 5]
```diff
@@ -685,6 +707,57 @@ def test_init_with_custom_env(tmp_path_factory: TempPathFactory) -> None:
...
+        ],
+        cwd=dummy_template["cwd"],
+        input="y\n",
+    )
+
+    assert result.exit_code == 1
+    verify(result.output, scrubber=make_output_scrubber())
+
+
+def test_init_fullstack_template_works(which_mock: WhichMock, dummy_template: dict[str, Path]) -> None:
```
Comment (neilcampbell) [inline tests/init/test_init.py:739]:
```suggestion
def test_init_template_with_python_task_works(which_mock: WhichMock, dummy_template: dict[str, Path]) -> None:
```

File: .github/actions/build-binaries/action.yaml [hunk 6]
```diff
@@ -40,22 +34,31 @@ runs:
...
+    - name: Upload binary as artifact (release)
       if: ${{ inputs.upload_binaries == 'true' }}
       uses: actions/upload-artifact@v4
       with:
-        name: algokit-cli-${{ inputs.operating_system }}-py${{ inputs.python_version }}
+        name: algokit-${{ inputs.operating_system }}-py${{ inputs.python_version }}
         path: algokit-${{ env.RELEASE_VERSION }}-${{ inputs.operating_system }}-py${{ inputs.python_version }}.tar.gz
 
+    - name: Upload binary as artifact (dev)
+      if: ${{ inputs.upload_binaries != 'true' }}
```
Comment (neilcampbell) [inline .github/actions/build-binaries/action.yaml:56]: @aorumbayev Should this input be renamed to something else?
Also should we be flagging these artefacts as pre-release?

**Review by aorumbayev — APPROVED** [review]
@neilcampbell approving Negar's bits but lmk if you have any further comments for the portability tests changes

File: .github/actions/build-binaries/action.yaml [hunk 7]
```diff
@@ -24,41 +24,61 @@ runs:
       run: ${{ inputs.build_command }}
       shell: bash
 
-    - name: Test Executable
+    - name: Set TEST_CLI_PATH
       run: |
-        ls -l dist
-        ./dist/${{ inputs.package_name }}/${{ inputs.package_name }}${{ inputs.operating_system == 'windows-latest' && '.exe' || '' }} --help
+        ls -la dist/algokit/
```
Comment (neilcampbell) [inline .github/actions/build-binaries/action.yaml:29]: Any reason to keep this?

File: src/algokit/cli/init.py (hunk 3 above)
Comment (neilcampbell) [reply src/algokit/cli/init.py:220]: @negar-abbasi Unmarked this as resolved, as it doesn't appear like this has been handled.

Comment (negar-abbasi) [reply src/algokit/cli/init.py:220]: I thought we are handling it with docs, telling user if they need this option then they have to do use 'python_path' convention as we did.

Comment (neilcampbell) [reply src/algokit/cli/init.py:220]: But we won't be resolving the `python_path` inside the `generate` command, so it won't work inside a generator.

File: .github/actions/build-binaries/action.yaml (hunk 7 above)
Comment (aorumbayev) [reply .github/actions/build-binaries/action.yaml]: @neilcampbell was useful for debugging but not anymore, can always revert back if needed

File: src/algokit/cli/init.py (hunk 3 above)
Comment (aorumbayev) [reply src/algokit/cli/init.py:220]: @negar-abbasi @neilcampbell see my last commit that adds the same logic to generate (i am not covering other default answers yet only the python_path answer for now). And there is an extra snapshot test that runs generate without mocking and tests that there is non zero exit code.

File: tests/init/test_init.py [hunk 8]
```diff
@@ -714,10 +714,7 @@ def test_init_template_with_python_task_fails_on_missing_python(
     verify(result.output, scrubber=make_output_scrubber())
 
 
-def test_init_template_with_python_task_works(
-    which_mock: WhichMock, dummy_algokit_template_with_python_task: dict[str, Path]
-) -> None:
-    which_mock.remove("python")
```
Comment (aorumbayev) [inline tests/init/test_init.py]: @negar-abbasi i haven't seen any effects of removing this i assume this was just a typo?

**Review by neilcampbell — APPROVED** [review]

File: tests/init/test_init.py (hunk 8 above)
Comment (aorumbayev) [reply tests/init/test_init.py]: @negar-abbasi i saw you restored the change back - can you explain why we need which_mock.remove('python') bit in this test?

File: src/algokit/cli/init.py (hunk 1 above)
Comment (achidlow) [reply src/algokit/cli/init.py:221]: ah, we need to update that to return `None` if `sys.executable and sys.executable.endswith("alogkit")` or something like that, it's not a valid interpreter so it shouldn't be returned. This goes for templates here as well as where it's used for `pipx`
